Feature/tour default popup

CloudNiner opened the pull request:
Add default detail popup on each tour view
![screen shot 2015-05-14 at 5 34 46 pm](https://cloud.githubusercontent.com/assets/1818302/7642510/944dd490-fa5f-11e4-8d32-c0c3744b7bf4.png)

CloudNiner committed:
Store/retrieve created visualizations via service

CloudNiner committed:
Display tour popup on cartodb visualization

flibbertigibbet commented on app/scripts/cartodb/vis-service.js:
```diff
@@ -0,0 +1,57 @@
+         * @param  {string} visId Visualization ID to retrieve
+         * @return {$q.deferred} Promise that resolves once visualization is ready.
+         *                               First argument is vis
+         *                               Second argument is layers
```
Second argument?

flibbertigibbet replied:
Oh, second argument to the returned promise, ok.

CloudNiner replied:
K, will clarify.

flibbertigibbet commented on app/scripts/views/tour/tour-page-controller.js:
```diff
@@ -5,24 +5,29 @@
+                                Config, TourConfig, TourPopupConfig, Visualization) {
 
+        var vis = null;
         var popupContent = '' +
```
`join` the template bits here

flibbertigibbet commented on app/scripts/views/tour/tour-controller.js:
```diff
@@ -5,12 +5,19 @@
 
         function initialize() {
-            cartodb.createVis('map', 'http://greenworks.cartodb.com/api/v2/viz/33627780-f988-11e4-8ff7-0e0c41326911/viz.json');
+            var vis = cartodb.createVis('map', 'http://greenworks.cartodb.com/api/v2/viz/33627780-f988-11e4-8ff7-0e0c41326911/viz.json');
```
Should let `vis-service` do the creation, maybe in `add`.

flibbertigibbet commented on app/scripts/views/tour/tour-controller.js:
```diff
@@ -5,12 +5,19 @@
+            // Delete vis on scope destroy to free resources
+            // and allow for it to be created again later
+            $scope.$on('$destroy', function () {
+                Visualization.remove(Config.ids.tour);
```
Seems like something more should happen than removing it from the collection of visualizations.  Will this remove it from the map?  Does it destroy the map?

CloudNiner replied:
You'd think so, but there is no API call in cartodbjs to destroy a visualization. The use case here is just to remove the objects, etc in js land when the view containing the map is destroyed. I can add some comments to this effect here.

flibbertigibbet commented on app/scripts/views/tour/tour-page-controller.js:
```diff
@@ -42,6 +47,15 @@
             ctl.onPrevClick = onPrevClick;
             ctl.onNextClick = onNextClick;
+
+            Visualization.get(Config.ids.tour).then(onVisReady, onVisError);
```
I understand wanting a separate manager for the visualizations, and separating controller and page logic, but this seems to be complicating matters, particularly with throwing an error for an expected condition in the visualization `get`.

An alternative suggestion:  maybe pass the viz ready/error callbacks to `add`, let `add` create the viz and invoke the callbacks when creation is `done`.  Don't add the viz ID to the collection until creation is done, and let `get` be a no-op until it's ready.

CloudNiner replied:
Hmmm, that may work. The root problem here though is that we need to create the visualization once (in the parent TourController), but then access it in child views multiple times. So either way, get needs to be a promise, because we don't know when we call get in the child controller whether the vis is fully initialized yet or not. And if I moved the callbacks to add, its superfluous because in the current implementation I can just create callbacks on the return value from add anyways (since its the cartodb vis object). 

Not sure what you're referring to when you say 'throwing an error for an expected condition'? 

Hope this clarifies things? I actually think it the current state its a relatively consistent API. The reason this is exists is so the visualization object can be passed around via a service rather than attaching it to the TourController and then accessing it implicitly in the child view, which is far more brittle.

Either way, I'll play with this a bit yet and see if I can come up with something a bit simpler.

flibbertigibbet replied:
The catch for `getLayers` failing if the layers aren't there yet is the expected condition (initialization not having finished yet) that I'm talking about; that sort of thing is generally a code smell.  Yes, I understand you're doing this to have it only created once.  What I'm suggesting is that if you pass in `onVisReady` and `onVisError` to `add`, you don't need to do the initialization work with `get`. `get` can be used then only to retrieve an already-initialized object you want to reference elsewhere.

flibbertigibbet commented:
+1

CloudNiner committed:
Cleanup jshint

CloudNiner commented:
Added additional documentation in README and vis-service.js after an offline discussion of how to handle this, so that it is a bit clearer as to what is going on there and why. 

Also added an additional check in the get method to ensure we're handling the error we expect.
